### PR TITLE
Implementation of the pick/set/incr/decr methods in CHAI

### PR DESCRIPTION
**Pull request by rfhaque**

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
Looks like the tests are failing:
```
3: /home/travis/build/LLNL/CHAI/src/tests/managed_array_tests.cpp:133: Failure
3:       Expected: temp
3:       Which is: 119
3: To be equal to: 5
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -49,6 +49,7 @@ set(ENABLE_CUDA On CACHE BOOL "Enable CUDA")
 set(ENABLE_OPENMP On CACHE BOOL "Enable OpenMP")
 set(ENABLE_BENCHMARKS On CACHE Bool "Enable benchmarks")
 option(ENABLE_CNMEM "Enable cnemem for GPU allocation" Off)
+option(ENABLE_PICK_SET_INCR_DECR "Enable pick/set/incr/decr" On)
```
Comment (davidbeckingsale) [inline CMakeLists.txt]: We can probably rename this option to `ENABLE_PICK` to keep it shorter.

File: cmake/thirdparty/SetupChaiThirdparty.cmake [hunk 2]
```diff
@@ -42,6 +42,7 @@
 #######################################################################
 
 message (STATUS "cnmem support is ${ENABLE_CNMEM}")
+message (STATUS "pick/set/incr/decr support is ${ENABLE_PICK_SET_INCR_DECR}")
```
Comment (davidbeckingsale) [inline cmake/thirdparty/SetupChaiThirdparty.cmake]: This message shouldn't be here, this is for third-party library stuff. Just delete the message.

File: src/ArrayManager.hpp [hunk 3]
```diff
@@ -50,6 +50,14 @@
 
 namespace chai {
 
+#if defined(CHAI_ENABLE_PICK_SET_INCR_DECR)
+#define DTOH 1
+#define HTOD 0
+
+#define __DTOH__(X) X == DTOH
```
Comment (davidbeckingsale) [inline src/ArrayManager.hpp]: I think having an enum for direction is better than macros. You can even use an enum class.

File: src/ArrayManager.hpp [hunk 4]
```diff
@@ -172,6 +180,85 @@ class ArrayManager
    */
   void free(void* pointer);
 
+#if defined(CHAI_ENABLE_PICK_SET_INCR_DECR)
```
Comment (davidbeckingsale) [inline src/ArrayManager.hpp]: Rename option.

File: src/ManagedArray.hpp [hunk 5]
```diff
@@ -156,6 +156,51 @@ class ManagedArray {
...
+  CHAI_HOST_DEVICE void pick(size_t i, T& val) const; 
+
+  /*!
+   * \brief Return the value of element i in the ManagedArray.
+   *
+   * \param index The index of the element to be fetched
+   * \return The value of the i-th element in the ManagedArray.
+   * \tparam T The type of data value in ManagedArray.
+   */
+  CHAI_HOST_DEVICE T pick(size_t i) const;
```
Comment (davidbeckingsale) [inline src/ManagedArray.hpp]: Is having two `pick` functions necessary in the interface?

File: src/ManagedArray.inl [hunk 6]
```diff
@@ -167,17 +167,63 @@ CHAI_HOST_DEVICE T& ManagedArray<T>::operator[](const int i) const {
...
+     //Executing on DEVICE. Simply dereference the active pointer 
+     ++m_active_pointer[i]; 
+  #endif
+}
+
+template<typename T>
+CHAI_INLINE
+CHAI_HOST_DEVICE void ManagedArray<T>::decr(size_t i) const { 
+  #if !defined(__CUDA_ARCH__)
+     //Executing on HOST. Decrement the value in the currently active execution space
```
Comment (davidbeckingsale) [inline src/ManagedArray.inl]: I think these comments are kind of redundant.

File: src/tests/managed_array_tests.cpp [hunk 7]
```diff
@@ -120,7 +120,192 @@ TEST(ManagedArray, Const) {
...
+  int temp;
+  temp = array.pick(5);
+  ASSERT_EQ(temp, 6);
+
+  temp = array.pick(9);
+  ASSERT_EQ(temp, 8);
+
+  array.free();
+}
+#endif
```
Comment (davidbeckingsale) [inline src/tests/managed_array_tests.cpp]: Can you add some UM `pick` tests please.

File: src/ManagedArray_thin.inl [hunk 8]
```diff
@@ -162,6 +162,52 @@ CHAI_HOST_DEVICE T& ManagedArray<T>::operator[](const int i) const {
...
+#if !defined(__CUDA_ARCH__) && defined(CHAI_ENABLE_UM)
+  cudaDeviceSynchronize();
+#endif
+  val = m_active_pointer[i]; 
+}
+
+template<typename T>
+CHAI_INLINE
+CHAI_HOST_DEVICE T ManagedArray<T>::pick(size_t i) const {
+  typename std::remove_const<T>::type temp;
```
Comment (davidbeckingsale) [inline src/ManagedArray_thin.inl]: We already have a `using` statement for the non-const T (in ManagedArray.hpp): 
```
using T_non_const = typename std::remove_const<T>::type;
```

**Comment by rfhaque** [comment]
Yes. Fixed the remove_const thing.

**Pull request merged**